x-pack/legacy/plugins/security/server/lib/authentication/authentication_result.ts

```diff
@@ -62,42 +62,58 @@ export class AuthenticationResult {
   /**
    * Produces `AuthenticationResult` for the case when authentication succeeds.
    * @param user User information retrieved as a result of successful authentication attempt.
+   * @param [authResponseHeaders] Optional dictionary of the HTTP headers with authentication
+   * information that should be specified in the response we send to the client request.
    * @param [state] Optional state to be stored and reused for the next request.
    */
-  public static succeeded(user: AuthenticatedUser, state?: unknown) {
+  public static succeeded(
+    user: AuthenticatedUser,
+    {
+      authResponseHeaders,
+      state,
+    }: Pick<AuthenticationOptions, 'authResponseHeaders' | 'state'> = {}
+  ) {
     if (!user) {
       throw new Error('User should be specified.');
     }
 
-    return new AuthenticationResult(AuthenticationResultStatus.Succeeded, { user, state });
+    return new AuthenticationResult(AuthenticationResultStatus.Succeeded, {
+      user,
+      authResponseHeaders,
+      state,
+    });
   }
 
   /**
    * Produces `AuthenticationResult` for the case when authentication fails.
    * @param error Error that occurred during authentication attempt.
-   * @param [challenges] Optional list of the challenges that will be returned to the user within
-   * `WWW-Authenticate` HTTP header. Multiple challenges will result in multiple headers (one per
-   * challenge) as it's better supported by the browsers than comma separated list within a single
-   * header. Challenges can only be set for errors with `401` error status.
-   */
-  public static failed(error: Error, challenges?: string[]) {
+   * @param [authResponseHeaders] Optional dictionary of the HTTP headers with authentication related
+   * information (e.g. `WWW-Authenticate` with the challenges) that should be specified in the
+   * response we send to the client request.
+   */
+  public static failed(
```

Comment:
__note:__ since we use response headers for not only `error` case we can generalize this code a bit and replace `challenges` with more generic `authResponseHeaders`.